### ISSUE TITLE
Understanding the loss functions

### ISSUE THREAD

**Issue by ErinYang96** [Kamlesh364/Mask-RCNN-TF2.7.0-keras2.7.0#7]
Hi,

I am checking my training results which are automatically recorded by `logs`. However, the current version of Mask-RCNN-TF2.7.0-keras2.7.0 includes more loss functions than the original version distributed by the Matterport Team, some of which is beyond my knowledge. I execute the command `tensorboard --inspect --logdir logs/trout20220603T1144`, and get outputs as follows.
```
Found event files in:
logs/trout20220603T1144

These tags are in logs/trout20220603T1144:
audio -
histograms -
images -
scalars
   epoch_loss
   epoch_mrcnn_bbox_loss
   epoch_mrcnn_class_loss
   epoch_mrcnn_mask_loss
   epoch_rpn_bbox_loss
   epoch_rpn_class_loss
   epoch_val_loss
   epoch_val_mrcnn_bbox_loss
   epoch_val_mrcnn_class_loss
   epoch_val_mrcnn_mask_loss
   epoch_val_rpn_bbox_loss
   epoch_val_rpn_class_loss
```
I cannot understand these loss functions very well cause I am quite new to deep learning frameworks. So could you please be so kind to briefly explain how to understand the functions listed above? Maybe other people who implement your code can also understand better with your simple guidance.
In my case, I set the epoch for head layer as 15, and epoch for all layers at 30. The results displayed by tensorboard is shown below.
![image](https://user-images.githubusercontent.com/100861190/174243727-7f6ef253-db7b-4893-8cb3-9d747c7ab479.png)
![image](https://user-images.githubusercontent.com/100861190/174243881-e8bb5bdb-53d4-4af2-aaec-058b6c8d8843.png)
![image](https://user-images.githubusercontent.com/100861190/174243986-0480280a-f07d-4084-832a-606e30c8789e.png)
![image](https://user-images.githubusercontent.com/100861190/174244670-18b2cd18-1ff4-43d9-9792-3e20a51bfe5e.png)
![image](https://user-images.githubusercontent.com/100861190/174244783-187c8a1e-68d4-4a6a-b0a5-2564d3cfd904.png)
![image](https://user-images.githubusercontent.com/100861190/174244908-edf40c42-5e00-41ed-b2ca-661e3cb78543.png)
![image](https://user-images.githubusercontent.com/100861190/174245002-44918c37-db21-4ac9-9937-96de7a520e57.png)
![image](https://user-images.githubusercontent.com/100861190/174245318-3179db7f-c815-47cf-8902-b3f9e72cc7c8.png)
![image](https://user-images.githubusercontent.com/100861190/174245506-d5b965bb-b3bc-4a17-beb5-7a9bb0bf214f.png)
![image](https://user-images.githubusercontent.com/100861190/174245637-1d5e67dc-a222-407d-b1be-5c72113164a3.png)
![image](https://user-images.githubusercontent.com/100861190/174245720-563f3472-02dd-4892-8eea-bb01f9dab770.png)
![image](https://user-images.githubusercontent.com/100861190/174245812-f13264f3-f0e9-482c-9024-fd6b52b3dbe3.png)

It would be great if you can take my case as an example for explanation. 
In addition, I have written my understanding of the original loss functions as follow (not sure if it is correct).
```
#### How to read the tensorboard results ####
  # Case 1: both loss() and val_loss() decline with epochs (best, the model is still learning)
  # Case 2: loss() declines, val_loss() fluctuates (overfit, maybe validation dataset is too close to training dataset)
  # Case 3: loss() fluctuates, val_loss() declines (something wrong with the datasets)
  # Case 4: both loss() and val_loss() fluctuate (reach the bottleneck, maybe you need to reduce the learning rate or batch size)
  # Case 5: both loss() and val_loss() increase (worst, maybe due to unsuitable network design or inappropriate hyperparameters setup)
```
If possible, please also give a brief guidance for reading the 12 loss functions (which trends denote good training, and which are not; if not, how to make changes on parameter setups).
This is a long question and a big favor to ask. Thanks first for your patience to read it, and also further thanks for your kind support in advance!

Best regards,
Erin

**Comment by prince0310** [comment]
Thanks for your question Erin , i have tried to discuss brief about all losses may be helpful for you
 
- epoch_loss -- The loss metric given in the log, is sum of other five losses
- epoch_mrcnn_bbox_loss -- Loss for Mask R-cnn bounding box refinement
- epoch_mrcnn_class_loss --  Loss for the classifier of mask R-cnn
- epoch_mrcnn_mask_loss -- Mask binary cross entropy loss for the mask 
- epoch_rpn_bbox_loss -- Rpn bounding box loss
- epoch_rpn_class_loss -- Rpn anchor classifier loss

-  Classification loss reflect the confidence of model that how the model predicting the particular class, rather we would say how closely model is predicting correct class.  **mrcnn_class_loss**: covers all objects classes but in **rpn_class_loss** only the classification that is done in labelling the anchor boxes as foreground and  background. i.e why the loss tends to have lower values.

- The bounding box loss indicates the distance between true box, dimension i.e width, height and (x,y) co-ordinates and predicted box  location. Basically it shows the model locating objects, how well within image. In case of **rpn_bbox_loss**, how good the model is at precisely predicting the area within image corresponding to the different objects that are present in the case of **mrcnn_bbox_loss**.

- Mask loss is similar to classification loss. Penalizes wrong per-pixels, binary classifications with respect to True class. Mask RCNN encodes each region of interest with binary mask and loss is calculated differently for each of the ROIs. 

**similar concept can be aplicable in validation set .** 
But Fluctuation in the validation loss can occur for various region. and it is difficult to guess at first sight. they might be caused by a learning rate or a validation set that is too small. 

Best regards,
Prince

**Comment by ErinYang96** [comment]
@prince0310 
Thanks so much for your explanations! The meanings of these functions are clearer now with your instructions.
